prov/shm/src/smr_progress.c

```diff
@@ -619,9 +619,25 @@ static void smr_progress_connreq(struct smr_ep *ep, struct smr_cmd *cmd)
 			"Error processing mapping request\n");
 
 	peer_smr = smr_peer_region(ep->region, idx);
+	assert(peer_smr != ep->region);
+	/*
+	 * we are currently under ep->region->lock.
+	 * To avoid dead lock, we need to release it
+	 * before acquiring peer lock. This is because in
+	 * two way communication both end points will
+	 * send and receive connreq and will try update peer.
+	 */
+	fastlock_release(&ep->region->lock);
```

Comment:
Is it safe to release this lock for a short duration, then re-acquire it?

Reply:
I believe it is a safe, not because duration is short, but because  no operation is performed on ```ep->region``` when lock is released.

Reply:
I was thinking about this as well. I'm not sure if it is. Another thread could come and try to progress the region and progress the same connreq command and add it into the smr_map twice. I'm still walking through the code to see what would happen but I think that could be an issue.

Reply:
Actually that wouldn't be an issue. The command queue is protected locally by the rx_cq lock. So I think it's ok. I'll double check the locking though.

Reply:
> progress the same connreq command and add it into the smr_map twice.

I think smr_map has its own lock to prevent that from happening, no?

Reply:
Ah yes, you're totally right. That would definitely protect it. I think the locking is fine then.

Reply:
Please do something different here.  Requiring that the caller acquire a lock, which this function releases, then re-acquires, is non-intuitive behavior and likely to lead to issues later.  It's particularly confusing having the tryacquire lock that now aborts the rest of the function, which was previously assumed to have completed.

Reply:
@shefty  I think you are right. The current PR is problematic. After the lock is released, another thread can come in and grab the same command, and we will have two threads processing same command and can both remove a command from command queue, resulting command leak. I will close the PR.

Reply:
I don't think that's the issue here since the commands are protected by rx_cq lock which isn't released in order to progress the commands. I think there actually is the possibility of a deadlock if the second thread acquires the ep lock and then hangs on acquiring the rx_cq lock and the first thread then can't re-acquire the ep lock.

Reply:
I see. thanks! Yeah, that make sense.